java/ffi/src/main/java/sleeper/foreign/FFIBridge.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.foreign;
+
+import jnr.ffi.LibraryLoader;
+import org.scijava.nativelib.JniExtractor;
+import org.scijava.nativelib.NativeLoader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Provides low-level bridge functionality for calling foreign code.
+ *
+ * @see FFIBridge#createForeignInterface(Class)
+ */
+public class FFIBridge {
+    /**
+     * Native library extraction object. This can extract native libraries from the classpath and
+     * unpack them to a temporary directory.
+     */
+    private static final JniExtractor EXTRACTOR = NativeLoader.getJniExtractor();
+
+    /** Paths in the JAR file where a native library may have been placed. */
+    private static final String[] LIB_PATHS = {
+        "natives/x86_64-unknown-linux-gnu/release", "natives/aarch64-unknown-linux-gnu/release",
+        // Rust debug builds will place libraries in different locations
+        "natives/x86_64-unknown-linux-gnu/debug", "natives/aarch64-unknown-linux-gnu/debug"};
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FFIBridge.class);
+
+    /**
+     * Attempt to load the native foreign function library.
+     *
+     * The native library will be extracted from the classpath and unpacked to a temporary
+     * directory. The library is then loaded and linked. Multiple locations are checked in the
+     * classpath, representing different architectures. Thus, if we attempt to load a library for
+     * the wrong CPU architecture, loading will fail and the next path will be tried. This way, we
+     * maintain a single JAR file that can work across multiple CPU architectures.
+     *
+     * Each call to this library will extract and link a new copy of the foreign library so clients
+     * are encouraged to cache results of successful calls to this function.
+     *
+     * @param  clazz       the interface describing the foreign function calls
+     * @param  <T>         interface type containing Java functions stubs
+     * @return             the native call interface
+     * @throws IOException if an error occurs during loading or linking the native library
+     */
+    public static synchronized <T extends ForeignFunctions> T createForeignInterface(Class<T> clazz) throws IOException {
+        try {
+            return extractAndLink(clazz, "compaction");
```

Comment:
This is passing in the library name specific to compaction. Should this be moved to the compaction module?

Reply:
Library renamed in #5299